File: android/src/main/kotlin/com/am/am_flutter_paypal/AmFlutterPaypalPlugin.kt
package com.am.am_flutter_paypal

import android.app.Activity
import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

import com.paypal.android.cardpayments.CardClient
import com.paypal.android.cardpayments.CardRequest
import com.paypal.android.cardpayments.Card
import com.paypal.android.corepayments.CoreConfig
import com.paypal.android.corepayments.Environment
import com.paypal.android.cardpayments.CardApproveOrderResult
import com.paypal.android.corepayments.PayPalSDKError

/** AmFlutterPaypalPlugin */
class AmFlutterPaypalPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {

    private lateinit var channel: MethodChannel
    private var activity: Activity? = null
    private var cardClient: CardClient? = null
    private val mainHandler = Handler(Looper.getMainLooper())
    private var returnUrl: String? = null
    private var coreConfig: CoreConfig? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "am_flutter_paypal")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "initialize" -> {
                val clientId = call.argument<String>("clientId")
                val environmentStr = call.argument<String>("environment")
                returnUrl = call.argument<String>("returnUrl")
                
                if (clientId == null || environmentStr == null) {
                    result.error("INVALID_ARGUMENTS", "ClientId or Environment is null", null)
                    return
                }

                val environment = if (environmentStr.equals("live", ignoreCase = true)) {
                    Environment.LIVE
                } else {
                    Environment.SANDBOX
                }

                coreConfig = CoreConfig(clientId, environment = environment)
                
                activity?.let {
                    cardClient = CardClient(it, coreConfig!!)
                    result.success(true)
                } ?: run {
                    // We can still initialize coreConfig and wait for activity
                    result.success(true)
                }
            }

            "approveOrder" -> {
                val orderId = call.argument<String>("orderId")
                val cardMap = call.argument<Map<String, Any>>("card")

                if (orderId == null || cardMap == null) {
                    result.error("INVALID_ARGUMENTS", "OrderId or Card details missing", null)
                    return
                }

                val currentActivity = activity
                if (currentActivity == null) {
                    result.error("ACTIVITY_NULL", "Activity is not attached", null)
                    return
                }

                val config = coreConfig
                if (config == null) {
                    result.error("SDK_NOT_INITIALIZED", "Plugin not initialized (Config null)", null)
                    return
                }

                // Initialize client if not already done
                if (cardClient == null) {
                    cardClient = CardClient(currentActivity, config)
                }

                val returnUrlSafe = returnUrl
                if (returnUrlSafe == null) {
                    result.error("MISSING_RETURN_URL", "Return URL must be provided during initialization", null)
                    return
                }

                val card = Card(
                    number = cardMap["cardNumber"] as String,
                    expirationMonth = cardMap["expirationMonth"] as String,
                    expirationYear = cardMap["expirationYear"] as String,
                    securityCode = cardMap["securityCode"] as String
                )
                card.cardholderName = cardMap["cardholderName"] as? String ?: ""

                val cardRequest = CardRequest(orderId, card, returnUrlSafe)
                
                cardClient?.approveOrder(cardRequest) { approveOrderResult ->
                    when (approveOrderResult) {
                        is CardApproveOrderResult.Success -> {
                            val response = mapOf(
                                "success" to true,
                                "orderId" to approveOrderResult.orderId,
                                "status" to "APPROVED"
                            )
                            send(result, response)
                        }
                        is CardApproveOrderResult.Failure -> {
                            val error = approveOrderResult.error
                            val response = mapOf(
                                "success" to false,
                                "errorCode" to error.code.toString(),
                                "message" to (error.errorDescription ?: "Unknown error")
                            )
                            send(result, response)
                        }
                        is CardApproveOrderResult.AuthorizationRequired -> {
                            // This state is for 3DS or other required actions
                            val response = mapOf(
                                "success" to false,
                                "errorCode" to "AUTHORIZATION_REQUIRED",
                                "message" to "Additional authorization or 3DS is required."
                            )
                            send(result, response)
                        }
                    }
                }
            }
            else -> result.notImplemented()
        }
    }

    private fun send(result: Result, data: Map<String, Any>) {
        mainHandler.post {
            result.success(data)
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        // Re-initialize client if config already exists
        coreConfig?.let {
            cardClient = CardClient(binding.activity, it)
        }
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
        cardClient = null
    }
}

File: ios/Classes/AmFlutterPaypalPlugin.swift
import Flutter
import UIKit
import PayPalCardPayments
import PayPalCorePayments

public class AmFlutterPaypalPlugin: NSObject, FlutterPlugin {
    private var cardClient: CardClient?
    private var channel: FlutterMethodChannel

    init(channel: FlutterMethodChannel) {
        self.channel = channel
        super.init()
    }

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "am_flutter_paypal", binaryMessenger: registrar.messenger())
        let instance = AmFlutterPaypalPlugin(channel: channel)
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialize":
            guard let args = call.arguments as? [String: Any],
                  let clientId = args["clientId"] as? String,
                  let environmentStr = args["environment"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "ClientId or Environment is missing", details: nil))
                return
            }

            let environment: Environment = (environmentStr.lowercased() == "live") ? .live : .sandbox
            let config = CoreConfig(clientID: clientId, environment: environment)
            self.cardClient = CardClient(config: config)
            // Note: returnUrl is mainly used on Android SDK 2.x, 
            // on iOS it's traditionally handled via the App's URL Scheme.
            result(true)

        case "approveOrder":
            guard let args = call.arguments as? [String: Any],
                  let orderId = args["orderId"] as? String,
                  let cardMap = args["card"] as? [String: Any] else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "OrderId or Card details missing", details: nil))
                return
            }

            guard let client = self.cardClient else {
                result(FlutterError(code: "SDK_NOT_INITIALIZED", message: "Plugin not initialized", details: nil))
                return
            }

            let card = Card(
                number: cardMap["cardNumber"] as? String ?? "",
                expirationMonth: cardMap["expirationMonth"] as? String ?? "",
                expirationYear: cardMap["expirationYear"] as? String ?? "",
                securityCode: cardMap["securityCode"] as? String ?? ""
            )
            card.cardholderName = cardMap["cardholderName"] as? String

            let cardRequest = CardRequest(orderID: orderId, card: card)
            
            client.delegate = self
            
            // Start the approval flow
            client.approveOrder(request: cardRequest)
            
            // Note: We don't return here yet, the delegate handles it.
            // But we need to keep track of the result callback if needed.
            // In this simple implementation, we assume only one call at a time which we guard in Dart.
            self.pendingResult = result

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private var pendingResult: FlutterResult?

    private func sendResponse(_ response: [String: Any]) {
        DispatchQueue.main.async {
            self.pendingResult?(response)
            self.pendingResult = nil
        }
    }
}

extension AmFlutterPaypalPlugin: CardDelegate {
    public func card(_ cardClient: CardClient, didFinishWithResult result: CardResult) {
        var response: [String: Any] = [:]
        response["success"] = true
        response["orderId"] = result.orderID
        response["status"] = result.status
        // Additional metadata could be added here
        sendResponse(response)
    }

    public func card(_ cardClient: CardClient, didFinishWithError error: CoreSDKError) {
        var response: [String: Any] = [:]
        response["success"] = false
        response["errorCode"] = String(error.code)
        response["message"] = error.errorDescription
        sendResponse(response)
    }

    public func cardDidCancel(_ cardClient: CardClient) {
        var response: [String: Any] = [:]
        response["success"] = false
        response["errorCode"] = "USER_CANCELLED"
        response["message"] = "The user cancelled the payment flow."
        sendResponse(response)
    }

    public func cardThreeDSecureWillAppear(_ cardClient: CardClient) {
        // Optional: Notify Flutter that 3DS UI is appearing
    }

    public func cardThreeDSecureDidFinish(_ cardClient: CardClient) {
        // Optional: Notify Flutter that 3DS UI finished
    }
}
